File: cpp-sudoku/src/run_generator.cpp
#include "run_generator.hpp"
#include "generator.hpp"
#include "reducer.hpp"
#include "utils.hpp"
#include "data_dirs.hpp"
#include "exceptions.hpp"
#include <iostream>
#include <thread>
#include <future>
#include <chrono>
#include <map>
#include <algorithm>

namespace sudoku {

void RunGenerator::generateMany() {
    auto start = std::chrono::steady_clock::now();
    std::filesystem::path reducerFailedDir = DataDirs::REDUCER_FAILED_DIR / Utils::getCurrentTime();
    std::map<long, long> counts;
    
    Generator generator(9, 31);
    Reducer reducer;

    std::atomic<long> totalCounter{0};
    std::atomic<long> successCounter{0};
    std::atomic<long> openMin{LONG_MAX};
    
    for (int i = 0; i < 10; ++i) {
        std::cout << "Generating " << ++totalCounter << std::endl;
        
        auto generateFuture = std::async(std::launch::async, [&generator]() {
            return generator.generate();
        });
        
        try {
            auto result = generateFuture.wait_for(GENERATOR_TIMEOUT);
            if (result == std::future_status::timeout) {
                std::cout << "Failed to generate (timeout)" << std::endl;
                continue;
            }
            
            auto sudoku = generateFuture.get();
            long openCount = Utils::countOpen(sudoku);
            
            // Only try to reduce if the puzzle has a reasonable number of clues (not 81)
            if (openCount < 81) {
                std::cout << "Minimizing " << totalCounter.load() << std::endl;
                auto minimizeFuture = std::async(std::launch::async, [&reducer, sudoku]() {
                    return reducer.reduce(sudoku);
                });
                
                try {
                    result = minimizeFuture.wait_for(REDUCER_TIMEOUT);
                    if (result == std::future_status::timeout) {
                        std::cout << "Failed to reduce: " << openCount << std::endl;
                        std::filesystem::path failedFile = reducerFailedDir / 
                            (std::to_string(openCount) + "-" + Utils::getCurrentTime() + "-" + std::to_string(totalCounter.load()) + ".txt");
                        Utils::writeFile(failedFile, Utils::toString2D(sudoku));
                        counts[300]++;
                        continue;
                    }
                    
                    auto minimizedSudoku = minimizeFuture.get();
                    long newOpenCount = Utils::countOpen(minimizedSudoku);
                    if (newOpenCount != openCount) {
                        std::cout << "Minimized: " << openCount << " => " << newOpenCount << std::endl;
                        openCount = newOpenCount;
                        sudoku = minimizedSudoku;
                    }
                } catch (const std::exception& e) {
                    std::cout << "Failed to reduce: " << openCount << std::endl;
                    std::filesystem::path failedFile = reducerFailedDir / 
                        (std::to_string(openCount) + "-" + Utils::getCurrentTime() + "-" + std::to_string(totalCounter.load()) + ".txt");
                    Utils::writeFile(failedFile, Utils::toString2D(sudoku));
                    counts[300]++;
                    continue;
                }
            } else {
                std::cout << "Generated complete puzzle (81 clues), skipping reduction" << std::endl;
            }
            
            long newMin = openCount;
            long oldMin = openMin.load();
            while (newMin < oldMin && !openMin.compare_exchange_weak(oldMin, newMin)) {
                // Retry until we successfully update the minimum
            }
            
            successCounter++;
            long successPercentage = 100 * successCounter.load() / totalCounter.load();
            std::cout << "Generated: " << openCount << "/" << openMin.load() << " - " 
                      << successCounter.load() << "/" << successPercentage << "%" << std::endl;
            std::cout << Utils::toString2D(sudoku) << std::endl;
            
            std::filesystem::path readyFile = DataDirs::READY_DIR / (std::to_string(openCount) + ".txt");
            Utils::appendFile(readyFile, Utils::toString1D(sudoku) + "\n");
            std::cout << "------------------" << std::endl;
            
            if (openCount < 100) {
                counts[openCount]++;
            }
            
        } catch (const ComplexityLimitException& e) {
            std::cout << "Failed to generate (complexity limit)" << std::endl;
            counts[200]++;
        } catch (const std::exception& e) {
            std::cout << "Failed to generate" << std::endl;
            counts[100]++;
        }
    }

    std::cout << "Counts: ";
    for (const auto& [count, frequency] : counts) {
        std::cout << count << ":" << frequency << " ";
    }
    std::cout << std::endl;
    
    auto end = std::chrono::steady_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
    std::cout << "Time: " << duration.count() / 1000.0 << std::endl;
}

} // namespace sudoku

File: src/main/java/self/ed/main/RunGenerator.java
package self.ed.main;

import static java.lang.Long.MAX_VALUE;
import static java.lang.System.currentTimeMillis;
import static java.util.concurrent.Executors.newSingleThreadExecutor;
import static java.util.concurrent.TimeUnit.SECONDS;
import static java.util.stream.Collectors.counting;
import static java.util.stream.Collectors.groupingBy;

import static self.ed.main.DataDirs.READY_DIR;
import static self.ed.main.DataDirs.REDUCER_FAILED_DIR;
import static self.ed.util.FileUtils.appendFile;
import static self.ed.util.FileUtils.writeFile;
import static self.ed.util.SudokuUtils.countOpen;
import static self.ed.util.SudokuUtils.getCurrentTime;
import static self.ed.util.SudokuUtils.toString1D;
import static self.ed.util.SudokuUtils.toString2D;

import java.nio.file.Path;
import java.time.Duration;
import java.util.Map;
import java.util.TreeMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Future;
import java.util.concurrent.atomic.AtomicLong;
import java.util.function.Function;
import java.util.stream.Stream;

import org.apache.commons.lang3.exception.ExceptionUtils;

import self.ed.exception.ComplexityLimitException;
import self.ed.generator.Generator;
import self.ed.generator.Reducer;

public class RunGenerator {

  private static final Duration GENERATOR_TIMEOUT = Duration.ofSeconds(3);
  private static final Duration REDUCER_TIMEOUT = Duration.ofSeconds(10);

  public static void main(String[] args) {
    generateMany();
  }

  private static void generateMany() {
    long start = currentTimeMillis();
    int complexityGenerateLimit = 31;
    Path reducerFailedDir = REDUCER_FAILED_DIR.resolve(getCurrentTime());
    Map<Long, Long> counts;
    try (ExecutorService executor = newSingleThreadExecutor()) {
      Generator generator = new Generator(9);
      Reducer reducer = new Reducer();

      AtomicLong totalCounter = new AtomicLong();
      AtomicLong successCounter = new AtomicLong();
      AtomicLong openMin = new AtomicLong(MAX_VALUE);
      counts = Stream.generate(() -> {
            System.out.println("Generating " + totalCounter.incrementAndGet());
            Future<Integer[][]> generateFuture = executor.submit(() -> generator.generate(complexityGenerateLimit));
            try {
              Integer[][] result = generateFuture.get(GENERATOR_TIMEOUT.toSeconds(), SECONDS);
              Long openCount = countOpen(result);
              Integer[][] res = result;
              System.out.println("Minimizing " + totalCounter.get());
              Future<Integer[][]> minimizeFuture = executor.submit(() -> reducer.reduce(res));
              try {
                result = minimizeFuture.get(REDUCER_TIMEOUT.toSeconds(), SECONDS);
                Long newOpenCount = countOpen(result);
                if (!newOpenCount.equals(openCount)) {
                  System.out.println("Minimized: " + openCount + " => " + newOpenCount);
                  openCount = newOpenCount;
                }
              } catch (Exception e) {
                minimizeFuture.cancel(true);
                System.out.println("Failed to reduce: " + openCount);
                Path failedFile = reducerFailedDir.resolve(
                    openCount + "-" + getCurrentTime() + "-" + totalCounter.get() + ".txt");
                writeFile(failedFile.toFile(), toString2D(result));
                return 300L;
              }
              Long newMin = openCount;
              openMin.getAndUpdate(oldMin -> Math.min(oldMin, newMin));
              successCounter.incrementAndGet();
              long successPercentage = 100 * successCounter.get() / totalCounter.get();
              System.out.println(
                  "Generated: " + openCount + "/" + openMin.get() + " - " + successCounter.get() + "/" + successPercentage
                      + "%");
              System.out.println(toString2D(result));
              Path readyFile = READY_DIR.resolve(openCount + ".txt");
              appendFile(readyFile.toFile(), toString1D(result) + "\n");
              System.out.println("------------------");
              return openCount;
            } catch (Exception e) {
              generateFuture.cancel(true);
              return ExceptionUtils.indexOfType(e, ComplexityLimitException.class) > -1 ? 200L : 100L;
            }
          })
          //.filter(openCount -> openCount < 100)
          .limit(10)
          .collect(groupingBy(Function.identity(), TreeMap::new, counting()));
    }

    System.out.println(counts);
    System.out.println("Time: " + (currentTimeMillis() - start) / 1000d);
  }
}
